src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -426,25 +570,60 @@ private void ReadExtendedAttributesBlock(Stream archiveStream)
             // 4096 is a common max path length, and also the size field is 12 bytes long, which is under int.MaxValue.
             if (_size > int.MaxValue)
             {
-                throw new InvalidOperationException(string.Format(SR.TarSizeFieldTooLargeForExtendedAttribute, _typeFlag.ToString()));
+                throw new InvalidOperationException(string.Format(SR.TarSizeFieldTooLargeForEntryType, _typeFlag.ToString()));
             }
 
             byte[] buffer = new byte[(int)_size];
             archiveStream.ReadExactly(buffer);
 
+            _extendedAttributes = ReadExtendedAttributesFromBuffer(buffer, _name);
+        }
+
+        // Asynchronously collects the extended attributes found in the data section of a PAX entry of type 'x' or 'g'.
+        // Throws if end of stream is reached or if an attribute is malformed.
+        private static async ValueTask<Dictionary<string, string>?> ReadExtendedAttributesBlockAsync(Stream archiveStream, TarEntryType entryType, long size, string name, CancellationToken cancellationToken)
+        {
+            Debug.Assert(entryType is TarEntryType.ExtendedAttributes or TarEntryType.GlobalExtendedAttributes);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // It is not expected that the extended attributes data section will be longer than int.MaxValue, considering
+            // 4096 is a common max path length, and also the size field is 12 bytes long, which is under int.MaxValue.
+            if (size > int.MaxValue)
```

Comment:
iirc `int.MaxValue > Array.MaxLength` and since we allocate an array of given size, it would be better to check for max array size

```suggestion
            if (size > Array.MaxLength)
```